test/csuite/checkpoint_snapshot_race/main.c

```diff
@@ -0,0 +1,575 @@
+void parse_sleep_config(const char *name, char *config_str, SLEEP_CONFIG *cfg);
+void sleep_for_us(TEST_OPTS *opts, WT_RAND_STATE *rnd, SLEEP_CONFIG cfg);
+
+static int open_cursor_wrap(TEST_OPTS *, WT_SESSION *, const char *, const char *, WT_CURSOR **);
+
+/*
+ * main --
+ *     Test's entry point.
+ */
+int
+main(int argc, char *argv[])
+{
+    CHECKPOINT_RACE_OPTS *cr_opts, _cr_opts;
+    TEST_OPTS *opts, _opts;
+    WT_SESSION *session;
+    pthread_t create_thread, ckpt_thread, validate_thread;
+
+    cr_opts = &_cr_opts;
+    opts = &_opts;
+    memset(opts, 0, sizeof(*opts));
+    memset(cr_opts, 0, sizeof(*cr_opts));
+    cr_opts->opts = opts;
+
+    testutil_check(testutil_parse_opts(argc, argv, opts));
+    testutil_make_work_dir(opts->home);
+
+    parse_sleep_config("mid_insertion", opts->insertion_sleep_str, &cr_opts->mid_insertion);
+    parse_sleep_config("checkpoint_start", opts->checkpoint_delay_str, &cr_opts->checkpoint_start);
+
+    /* Default to 15 seconds */
+    if (opts->runtime == 0)
+        opts->runtime = 15;
+
+    /*
+     * Start collection counter at 10, since it's used as a proxy for timestamps as well which makes
+     * it better to avoid 0.
+     */
+    cr_opts->collection_count = 10;
+    testutil_check(pthread_mutex_init(&cr_opts->ckpt_go_cond_mutex, NULL));
+    testutil_check(pthread_cond_init(&cr_opts->ckpt_go_cond, NULL));
+
+    testutil_check(wiredtiger_open(opts->home, NULL,
+      "create,cache_size=100MB,log=(enabled,file_max=10M,remove=true),"
+      "debug_mode=(table_logging,eviction)",
+      &opts->conn));
+    testutil_check(opts->conn->open_session(opts->conn, NULL, NULL, &session));
+
+    /* Setup globaal transaction IDs. */
+    opts->conn->set_timestamp(opts->conn, "oldest_timestamp=1,stable_timestamp=1");
+
+    /* Create the catalogue table. */
+    testutil_check(
+      session->create(session, CATALOG_URI, "key_format=Q,value_format=SS,log=(enabled=false)"));
+    /* Create a table that is used to add load to the cache/database. */
+    testutil_check(
+      session->create(session, LOAD_TABLE_URI, "key_format=Q,value_format=SS,log=(enabled=false)"));
+
+    /* Spawn threads */
+    testutil_check(pthread_create(&ckpt_thread, NULL, thread_checkpoint, cr_opts));
+    testutil_check(pthread_create(&create_thread, NULL, thread_create_table_race, cr_opts));
+    testutil_check(pthread_create(&create_thread, NULL, thread_add_load, cr_opts));
+    testutil_check(pthread_create(&validate_thread, NULL, thread_validate, cr_opts));
+    usleep(200); /* Wait for threads to spin up */
+
+    snprintf(opts->progress_msg, opts->progress_msg_len, "Running for %" PRIu64 " seconds\n",
+      opts->runtime);
+    testutil_progress(opts, opts->progress_msg);
+    sleep((unsigned int)opts->runtime);
+    test_running = false;
+
+    testutil_progress(opts, "Stopping\n");
+    sleep(1);
+    testutil_check(pthread_join(ckpt_thread, NULL));
+    testutil_check(pthread_join(create_thread, NULL));
```

Comment:
You'd need to add another join for the additional thread. What I'm seeing is that all of the re-entrant type errors show this in the log:
```
[1654493138] Stopping

[1654493138] END validate thread, validation_passes: 0, validated_values: 0

[1654493139] END ckpt thread
```
Notice the missing one is the `END` message for the create_table_race function. I cannot exactly see how it generates the re-entrant error but it is a possible uncleanliness in the program that could be acting in weird ways.